sklearn/preprocessing/_discretization.py

```diff
@@ -231,32 +261,71 @@ def fit(self, X, y=None, sample_weight=None):
 
         n_samples, n_features = X.shape
 
-        if sample_weight is not None and self.strategy == "uniform":
-            raise ValueError(
-                "`sample_weight` was provided but it cannot be "
-                "used with strategy='uniform'. Got strategy="
-                f"{self.strategy!r} instead."
-            )
-
         if self.subsample is not None and n_samples > self.subsample:
             # Take a subsample of `X`
+            # When resampling, it is important to subsample **with replacement** to
+            # preserve the distribution, in particular in the presence of a few data
+            # points with large weights.
             X = resample(
                 X,
-                replace=False,
+                replace=True,
                 n_samples=self.subsample,
                 random_state=self.random_state,
+                sample_weight=sample_weight,
             )
-
-        n_features = X.shape[1]
-        n_bins = self._validate_n_bins(n_features)
+            # Since we already used the weights when resampling when provided,
+            # we set them back to `None` to avoid accounting for the weights twice
+            # in subsequent operations to compute weight-aware bin edges with
+            # quantiles or k-means.
+            sample_weight = None
 
         if sample_weight is not None:
             sample_weight = _check_sample_weight(sample_weight, X, dtype=X.dtype)
 
+        n_features = X.shape[1]
+        n_bins = self._validate_n_bins(n_features)
+
         bin_edges = np.zeros(n_features, dtype=object)
+
+        if self.strategy == "quantile" and self.quantile_method == "warn":
+            if sample_weight is None:
+                warnings.warn(
+                    "Defaulting to quantile method linear this will be changed to\
+                                  average_inverted_cdf in scikit-learn version 1.9",
+                    FutureWarning,
+                )
+                self.quantile_method = "linear"
+            if sample_weight is not None:
+                warnings.warn(
+                    "Defaulting to quantile method inverted_cdf this will be changed to\
+                                  average_inverted_cdf in scikit-learn version 1.9",
+                    FutureWarning,
+                )
+                self.quantile_method = "inverted_cdf"
+
+        if (
+            self.quantile_method not in ["inverted_cdf", "averaged_inverted_cdf"]
+            and sample_weight is not None
+        ):
+            raise ValueError(
+                "When using quantile strategy with sample weights, quantile method\
+                      should be inverted_cdf or averaged_inverted_cdf"
```

Comment:
Better recall the actual observed value of the parameter in the message to help the user understand what's going on when getting this error:

```suggestion
                "When fitting with strategy='quantile' and sample weights, "
                "quantile_method should either be set to 'averaged_inverted_cdf' or "
                f"'inverted_cdf', got quantile_method='{self.quantile_method}' instead".
```